services/audit/audit.go

```diff
@@ -0,0 +1,167 @@
+		return t.RemoteAddress
+	}
+
+	if d.Type == audit_model.TypeSystem {
+		return "System"
+	}
+
+	return ""
+}
+
+func (d TypeDescriptor) HTMLURL() string {
+	switch t := d.Object.(type) {
+	case *repository_model.Repository:
+		return t.HTMLURL()
+	case *user_model.User:
+		return t.HTMLURL()
+	case *organization_model.Organization:
+		return t.HTMLURL()
+	}
+	return ""
+}
+
+func Init() error {
+	if !setting.Audit.Enabled {
+		return nil
+	}
+
+	return initAuditFile()
+}
+
+var systemObject struct{}
+
+func scopeToDescription(scope any) TypeDescriptor {
+	if scope == &systemObject {
+		return TypeDescriptor{audit_model.TypeSystem, 0, nil}
+	}
+
+	switch s := scope.(type) {
+	case *repository_model.Repository, *user_model.User, *organization_model.Organization:
+		return typeToDescription(scope)
+	default:
+		panic(fmt.Sprintf("unsupported scope type: %T", s))
+	}
+}
+
+func typeToDescription(val any) TypeDescriptor {
+	if val == &systemObject {
+		return TypeDescriptor{audit_model.TypeSystem, 0, nil}
+	}
+
+	switch t := val.(type) {
+	case *repository_model.Repository:
+		return TypeDescriptor{audit_model.TypeRepository, t.ID, val}
+	case *user_model.User:
+		if t.IsOrganization() {
+			return TypeDescriptor{audit_model.TypeOrganization, t.ID, val}
+		}
+		return TypeDescriptor{audit_model.TypeUser, t.ID, val}
+	case *organization_model.Organization:
+		return TypeDescriptor{audit_model.TypeOrganization, t.ID, val}
+	case *user_model.EmailAddress:
+		return TypeDescriptor{audit_model.TypeEmailAddress, t.ID, val}
+	case *organization_model.Team:
+		return TypeDescriptor{audit_model.TypeTeam, t.ID, val}
+	case *auth_model.WebAuthnCredential:
+		return TypeDescriptor{audit_model.TypeWebAuthnCredential, t.ID, val}
+	case *user_model.UserOpenID:
+		return TypeDescriptor{audit_model.TypeOpenID, t.ID, val}
+	case *auth_model.AccessToken:
+		return TypeDescriptor{audit_model.TypeAccessToken, t.ID, val}
+	case *auth_model.OAuth2Application:
+		return TypeDescriptor{audit_model.TypeOAuth2Application, t.ID, val}
+	case *auth_model.Source:
+		return TypeDescriptor{audit_model.TypeAuthenticationSource, t.ID, val}
+	case *asymkey_model.PublicKey:
+		return TypeDescriptor{audit_model.TypePublicKey, t.ID, val}
+	case *asymkey_model.GPGKey:
+		return TypeDescriptor{audit_model.TypeGPGKey, t.ID, val}
+	case *secret_model.Secret:
+		return TypeDescriptor{audit_model.TypeSecret, t.ID, val}
+	case *webhook_model.Webhook:
+		return TypeDescriptor{audit_model.TypeWebhook, t.ID, val}
+	case *git_model.ProtectedTag:
+		return TypeDescriptor{audit_model.TypeProtectedTag, t.ID, val}
+	case *git_model.ProtectedBranch:
+		return TypeDescriptor{audit_model.TypeProtectedBranch, t.ID, val}
+	case *repository_model.PushMirror:
+		return TypeDescriptor{audit_model.TypePushMirror, t.ID, val}
+	default:
+		panic(fmt.Sprintf("unsupported type: %T", t))
+	}
+}
+
+func tryGetIPAddress(ctx context.Context) string {
+	if req := middleware.GetContextRequest(ctx); req != nil {
```

Comment:
How to handle the Gitea instances behind reverse proxies?

Reply:
It's the same variable we use everywhere else when we need the IP address. If the reverse proxy settings are correct, this should be the real address.